hs/app/reachpc/ReachPC/DirScan.hs

```diff
@@ -55,54 +68,85 @@ compareHashTrees localHashes remoteHashes = M.union updatedFiles deletedFiles
-        return ignores
-  doScan _ [] hashTree = return hashTree
-  doScan ignores (dir:dirStack) hashTree = do
+
+  doScan _ _ _ [] hashTree = return hashTree
+  doScan cacheTree cacheMTime ignores (dir:dirStack) hashTree = do
     -- Read .reachignore in this directory
     newIgnores <- readIgnores dir
     let ignores' = newIgnores <> ignores
     
     -- Walk current dir
-    paths <- listDirectory dir
-    let paths' = map (dir </>) paths
-    pathsAreFile <- mapM doesFileExist paths'
-    pathsAreSymlink <- mapM pathIsSymbolicLink paths'
-    let pathTypes = zip3 paths' pathsAreFile pathsAreSymlink
+    paths <- map (dir </>) <$> listDirectory dir
+    pathsAreFile <- mapM doesFileExist paths
+    pathsAreSymlink <- mapM pathIsSymbolicLink paths
+    let pathTypes = zip3 paths pathsAreFile pathsAreSymlink
     let pathTypes' = [(p, isFile) | (p, isFile, isSymlink) <- pathTypes, 
                                     not isSymlink, 
                                     not $ isIgnored ignores' p]
-    let (files, dirs) = partition snd pathTypes'
-    let files' = map fst files
-    let dirs' = map fst dirs
+    let (files_, dirs_) = partition snd pathTypes'
+    let files = map fst files_
+    let dirs = map fst dirs_
     
     -- Hash encountered files
-    fileHashes <- mapM hashFile files'
-    let fileHashes' = zip files' (map FH_File fileHashes)
+    let cacheOrHash path = do
+          mtime <- getModificationTime path
+          case mtime < cacheMTime of
+            False -> hashFile path
+            True -> case M.lookup path cacheTree of
+                      Just h -> return h
+                      _ -> hashFile path
+    fileHashes <- mapM cacheOrHash files
+    let fileHashes' = zip files (map FH_File fileHashes)
     let newFilesMap = M.fromList fileHashes'
 
     -- Add remaining dirs to stack and hashes map
-    let dirStack' = dirs' <> dirStack
-    let newDirsMap = M.fromList (map (, FH_Directory) dirs')
+    let dirStack' = dirs <> dirStack
+    let newDirsMap = M.fromList (map (, FH_Directory) dirs)
     
     let hashTree' = M.unions [hashTree, newFilesMap, newDirsMap]
-    doScan ignores' dirStack' hashTree'
+    doScan cacheTree cacheMTime ignores' dirStack' hashTree'
+
+  isIgnored ignores path = any (`G.match` path) ignores
+  readIgnores dir = do
+    let path = dir </> ".reachignore"
+    haveIgnores <- doesFileExist path
+    case haveIgnores of
+      False -> return []
+      True -> do
+        ignoresFile <- readFile path
+        let patternChar c = not $ isSpace c || c == '#'
+        let ignoresPatterns = filter (not . null) $ map (takeWhile patternChar) $ lines ignoresFile
+        let ignores = map (G.compile . (dir </>)) ignoresPatterns
+        return ignores
+
+  hashTreeToCacheTree = M.fromList . foldr foldHelper [] . M.toList
+  foldHelper (p, fh) lst =
+    case fh of
+      FH_File h -> (p, h) : lst
+      _ -> lst
+
+  cachePath = root </> ".reachcache"
```

Comment:
We already have `.reach` directory... I think it would be best to use it rather than have too many `.reach` files